src/test/socket/bind/test_bind.rs

```diff
@@ -0,0 +1,588 @@
+            for (test_name, _) in more_tests.iter_mut() {
+                *test_name = format!("{} <type={},flag={}>", test_name, sock_type, flag);
+            }
+
+            tests.extend(more_tests);
+        }
+    }
+
+    let num_tests = tests.len();
+    let mut tests: std::collections::BTreeMap<_, _> = tests.into_iter().collect();
+
+    // make sure we didn't have any duplicate tests
+    assert_eq!(num_tests, tests.len());
+
+    // add all of the passing tests
+    tests.extend(get_passing_tests());
+
+    tests
+}
+
+fn run_tests<'a, I>(tests: I, summarize: bool) -> Result<(), ()>
+where
+    I: Iterator<Item = (&'a String, &'a TestFn)>,
+{
+    for (test_name, test_fn) in tests {
+        print!("Testing {}...", test_name);
+
+        match test_fn() {
+            Err(msg) => {
+                println!(" ✗ ({})", msg);
+                if !summarize {
+                    return Err(());
+                }
+            }
+            Ok(_) => {
+                println!(" ✓");
+            }
+        }
+    }
+
+    Ok(())
+}
+
+// test binding using an argument that cannot be a fd
+fn test_invalid_fd() -> Result<(), String> {
+    let args = BindArguments {
+        fd: -1,
+        addr: None,
+        addr_len: Some(5),
+    };
+
+    check_bind_call(&args, true, Some(libc::EBADF))?;
+
+    Ok(())
+}
+
+// test binding using an argument that could be a fd, but is not
+fn test_non_existent_fd() -> Result<(), String> {
+    let args = BindArguments {
+        fd: 8934,
+        addr: None,
+        addr_len: Some(5),
+    };
+
+    check_bind_call(&args, true, Some(libc::EBADF))?;
+
+    Ok(())
+}
+
+// test binding a valid fd that is not a socket
+fn test_non_socket_fd() -> Result<(), String> {
+    let args = BindArguments {
+        fd: 0, // assume the fd 0 is already open and is not a socket
+        addr: None,
+        addr_len: Some(5),
+    };
+
+    check_bind_call(&args, true, Some(libc::ENOTSOCK))?;
+
+    Ok(())
+}
+
+// test binding a valid fd, but with a NULL address
+fn test_null_addr() -> Result<(), String> {
+    let fd = unsafe { libc::socket(libc::AF_INET, libc::SOCK_STREAM, 0) };
+    assert!(fd >= 0);
+
+    let args = BindArguments {
+        fd: fd,
+        addr: None,
+        addr_len: Some(5),
+    };
+
+    run_and_close_fds(&[fd], || check_bind_call(&args, true, Some(libc::EFAULT)))?;
+
+    Ok(())
+}
+
+// test binding a valid fd and address, but an address length that is too low
+fn test_short_addr() -> Result<(), String> {
+    let fd = unsafe { libc::socket(libc::AF_INET, libc::SOCK_STREAM, 0) };
+    assert!(fd >= 0);
+
+    let addr = libc::sockaddr_in {
+        sin_family: libc::AF_INET as u16,
+        sin_port: 11111u16.to_be(),
+        sin_addr: libc::in_addr {
+            s_addr: libc::INADDR_LOOPBACK.to_be(),
+        },
+        sin_zero: [0; 8],
+    };
+
+    let args = BindArguments {
+        fd: fd,
+        addr: Some(LibcSockAddr::In(addr)),
+        addr_len: Some(5),
```

Comment:
Maybe clearer as `size_of<libc::sockaddr_in>-1` instead of `5`?

Reply:
Done